src/ser.rs

```diff
@@ -436,26 +420,66 @@ impl<W: enc::Write> serde::ser::SerializeTupleVariant for BoundedCollect<'_, W>
     }
 }
 
-impl<W: enc::Write> serde::ser::SerializeMap for Collect<'_, W> {
+/// CBOR RFC-7049 specifies a canonical sort order, where keys are sorted by length first. This
+/// was later revised with RFC-8949, but we need to stick to the original order to stay compatible
+/// with existing data.
+/// We first serialize each map entry (the key and the value) into a buffer and then sort those
+/// buffers. Byte-wise comparison gives us the right order as keys in DAG-CBOR are always strings
+/// and prefixed with the length. Once sorted they are written to the actual output.
+pub struct CollectMap<'a, W> {
+    buffer: BufWriter,
+    entries: Vec<Vec<u8>>,
+    ser: &'a mut Serializer<W>,
+}
+
+impl<'a, W> CollectMap<'a, W> {
+    fn new(ser: &'a mut Serializer<W>) -> Self {
+        Self {
+            buffer: BufWriter::new(Vec::new()),
+            entries: Vec::new(),
+            ser,
+        }
+    }
+}
+
+impl<W> serde::ser::SerializeMap for CollectMap<'_, W>
+where
+    W: enc::Write,
+{
     type Ok = ();
     type Error = EncodeError<W::Error>;
 
     #[inline]
     fn serialize_key<T: Serialize + ?Sized>(&mut self, key: &T) -> Result<(), Self::Error> {
-        key.serialize(&mut *self.ser)
+        // Instantiate a new serializer, so that the buffer can be re-used.
+        let mut mem_serializer = Serializer::new(&mut self.buffer);
+        key.serialize(&mut mem_serializer)
+            .map_err(|_| EncodeError::Msg("Map key cannot be serialized.".to_string()))?;
+        Ok(())
     }
 
     #[inline]
     fn serialize_value<T: Serialize + ?Sized>(&mut self, value: &T) -> Result<(), Self::Error> {
-        value.serialize(&mut *self.ser)
+        // Instantiate a new serializer, so that the buffer can be re-used.
+        let mut mem_serializer = Serializer::new(&mut self.buffer);
+        value
+            .serialize(&mut mem_serializer)
+            .map_err(|_| EncodeError::Msg("Map value cannot be serialized.".to_string()))?;
+
+        self.entries.push(self.buffer.buffer().to_vec());
+        self.buffer.clear();
```

Comment:
(as an optimization, we can _could_ check "is sorted" and then just copy the entire buffer instead of sorting and copying slices... but that's not critical)

Reply:
Agreed, though it would make the sorting quite a bit more complicated. If someone  micro-benchmarks and implements it, I'd surely take it. BTW: this is exactly how it also works in the current version. This PR just moves it to a different location.